algorithms/a3c/env_vizdoom.py

```diff
@@ -0,0 +1,379 @@
+                                 x.name == self.objects[correct_object_id]][0]
+
+        # Special code to handle 'largest' and 'smallest' since we need to
+        # compute sizes for those particular instructions.
+        if 'largest' not in self.instruction \
+                and 'smallest' not in self.instruction:
+            object_ids = random.sample([x for x in range(len(self.objects))
+                                        if x not in correct_objects], 4)
+        else:
+            object_ids = self.get_candidate_objects_superlative_instr(
+                         chosen_correct_object)
+            object_ids = [self.object_dict[x] for x in object_ids]
+
+        assert len(object_ids) == 4
+
+        object_ids.insert(self.correct_location, correct_object_id)
+
+        # Get agent and object spawn locations.
+        agent_x_coordinate, agent_y_coordinate, \
+            agent_orientation, object_x_coordinates, \
+            object_y_coordinates = self.get_agent_and_object_positions()
+
+        self.object_coordinates = [ObjectLocation(x, y) for x, y in
+                                   zip(object_x_coordinates,
+                                       object_y_coordinates)]
+
+        # Spawn agent.
+        spawn_agent(self.game, agent_x_coordinate,
+                    agent_y_coordinate, agent_orientation)
+
+        # Spawn objects.
+        [spawn_object(self.game, object_id, pos_x, pos_y) for
+            object_id, pos_x, pos_y in
+            zip(object_ids, object_x_coordinates, object_y_coordinates)]
+
+        screen = self.game.get_state().screen_buffer
+        screen_buf = process_screen(screen, self.params.frame_height,
+                                    self.params.frame_width)
+
+        state = (screen_buf, self.instruction)
+
+        print('New instruction: {}'.format(self.instruction))
+
+        # this function used to return a tuple of size 4, get_reward() was calculated
+        # but this was removed to make it the same as OpenAI gym style for maximum flexibility
+        return state
+
+    def step(self, action_id):
+        """Executes an action in the environment to reach a new state.
+        Args:
+          action_id: An integer corresponding to the action.
+        Returns:
+           state: A tuple of screen buffer state and instruction.
+           reward: Reward at that step.
+           is_final: Flag indicating terminal state.
+           extra_args: Dictionary of additional arguments/parameters.
+        """
+        # Repeat the action for 5 frames.
+        if self.params.visualize:
+            # Render 5 frames for better visualization.
+            for _ in range(5):
+                self.game.make_action(actions[action_id], 1)
+                # Slowing down the game for better visualization.
+                sleep(self.params.sleep)
+        else:
+            self.game.make_action(actions[action_id], 5)
+
+        self.time += 1
+        reward = self.get_reward()
```

Comment:
I didn't write this ViZDoom code, so I'm not gonna change their env class to be like ai2thor env class, we just wanted it to work, not to rewrite it.

Reply:
It is not too much effort to modify that, but if you prefer it ugly then leave it as is.

Reply:
you're suggesting to put vizdoom code into our tasks.py. that's a big ask

Reply:
if you agree that we should use ViZDoom because of these reasons: 
1. We already have it installed
2. It's super customisable (check these tools)
3. Up to 7000 fps in sync mode, single threaded

Then I'll try to make our tasks.py work with it